diff -urN 4.0.11.60/CORE/HDD/src/wlan_hdd_cfg80211.c 4.0.11.60_ch/CORE/HDD/src/wlan_hdd_cfg80211.c
--- 4.0.11.60/CORE/HDD/src/wlan_hdd_cfg80211.c	2016-04-15 18:39:44.000000000 +0800
+++ 4.0.11.60_ch/CORE/HDD/src/wlan_hdd_cfg80211.c	2016-05-20 10:43:18.270600825 +0800
@@ -11848,15 +11848,6 @@
         return -ENOMEM;
     }
 
-    /* Extract and add the extended capabilities and interworking IE */
-    wlan_hdd_add_extra_ie(pHostapdAdapter, genie, &total_ielen,
-                          WLAN_EID_EXT_CAPABILITY);
-
-    wlan_hdd_add_extra_ie(pHostapdAdapter, genie, &total_ielen,
-                          WLAN_EID_INTERWORKING);
-
-    wlan_hdd_add_extra_ie(pHostapdAdapter, genie, &total_ielen,
-                          WLAN_EID_VHT_TX_POWER_ENVELOPE);
     if (0 != wlan_hdd_add_ie(pHostapdAdapter, genie,
                               &total_ielen, WPS_OUI_TYPE, WPS_OUI_TYPE_SIZE))
     {
