diff --git a/src/drivers/drivers.mak b/src/drivers/drivers.mak
index 68ff910..990ec85 100644
--- a/src/drivers/drivers.mak
+++ b/src/drivers/drivers.mak
@@ -28,9 +28,10 @@ NEED_LINUX_IOCTL=y
 NEED_RFKILL=y
 
 ifdef CONFIG_LIBNL32
-  DRV_LIBS += -lnl-3
-  DRV_LIBS += -lnl-genl-3
-  DRV_CFLAGS += -DCONFIG_LIBNL20 -I/usr/include/libnl3
+  DRV_LIBS += ${INCLUDE_LIBNL32_LIB}
+  DRV_LIBS += ${INCLUDE_LIBNL32_GENL_LIB}
+  DRV_LIBS += -lpthread
+  DRV_CFLAGS += -DCONFIG_LIBNL20 -I${INCLUDE_LIBNL32_HEADER}
 else
   ifdef CONFIG_LIBNL_TINY
     DRV_LIBS += -lnl-tiny
@@ -41,6 +42,7 @@ else
   ifdef CONFIG_LIBNL20
     DRV_LIBS += -lnl-genl
     DRV_CFLAGS += -DCONFIG_LIBNL20
+    DRV_LIBS += -lm -lpthread
   endif
 endif
 endif
@@ -153,9 +155,9 @@ endif
 ifdef CONFIG_VLAN_NETLINK
 ifdef CONFIG_FULL_DYNAMIC_VLAN
 ifdef CONFIG_LIBNL32
-  DRV_LIBS += -lnl-3
-  DRV_LIBS += -lnl-genl-3
-  DRV_LIBS += -lnl-route-3
+  DRV_LIBS += ${INCLUDE_LIBNL32_LIB}
+  DRV_LIBS += ${INCLUDE_LIBNL32_GENL_LIB}
+  DRV_LIBS += ${INCLUDE_LIBNL32_ROUTE_LIB}
   DRV_CFLAGS += -DCONFIG_LIBNL20
 else
   ifdef CONFIG_LIBNL_TINY
