diff -Nuar a/external/sepolicy/service_contexts b/external/sepolicy/service_contexts
--- a/external/sepolicy/service_contexts	2015-11-04 11:50:47.006410097 +0800
+++ b/external/sepolicy/service_contexts	2015-11-04 11:51:04.642409988 +0800
@@ -119,6 +119,7 @@
 wifip2p                                   u:object_r:system_server_service:s0
 wifiscanner                               u:object_r:system_server_service:s0
 wifi                                      u:object_r:system_server_service:s0
+qcomsoftap                                u:object_r:system_server_service:s0
 window                                    u:object_r:system_server_service:s0
 
 *                                         u:object_r:default_android_service:s0
