From 5480d849654d11c49af7c4e520718955432fc729 Mon Sep 17 00:00:00 2001
From: Yi Chen <yichen@qti.qualcomm.com>
Date: Tue, 26 Jan 2016 14:32:20 +0800
Subject: [PATCH 1/7] QCACLD Kbuild. Signed-off-by: Yi Chen
 <yichen@qti.qualcomm.com>

---
 Kbuild |   34 ++++++++++++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/Kbuild b/Kbuild
index 3f7238b..9af267d 100644
--- a/Kbuild
+++ b/Kbuild
@@ -399,6 +399,10 @@ ifeq ($(CONFIG_WLAN_FEATURE_MEMDUMP),y)
 HDD_OBJS += $(HDD_SRC_DIR)/wlan_hdd_memdump.o
 endif
 
+ifeq ($(CONFIG_DIRECT_AUDIO),y)
+HDD_OBJS +=	$(HDD_SRC_DIR)/wlan_hdd_direct_audio.o
+endif
+
 ############ EPPING ############
 EPPING_DIR :=	CORE/EPPING
 EPPING_INC_DIR :=	$(EPPING_DIR)/inc
@@ -1271,6 +1275,7 @@ endif
 #Enable GTK Offload
 ifeq ($(CONFIG_EXT_WOW), 1)
 CDEFINES += -DWLAN_FEATURE_EXTWOW_SUPPORT
+CDEFINES += -DWLAN_FEATURE_EXTWOW_OUTDOOR_SUPPORT
 endif
 
 #Mark it as SMP Kernel
@@ -1406,6 +1411,35 @@ ifeq ($(CONFIG_WLAN_SYNC_TSF),y)
 CDEFINES += -DWLAN_FEATURE_TSF
 endif
 
+ifeq ($(CONFIG_DIRECT_AUDIO),y)
+CDEFINES += -DDIRECT_AUDIO_SUPPORT
+ifeq ($(CONFIG_ROME_IF),usb)
+CDEFINES += -DCONFIG_CONTX_USB_DEBUG
+endif
+endif
+
+ifeq ($(CONFIG_VOS_MEM_PRE_ALLOC),y)
+CDEFINES += -DCONFIG_VOS_MEM_PRE_ALLOC
+endif
+
+ifeq ($(CONFIG_SS_REGULATORY),y)
+CDEFINES += -DSS_REGULATORY
+endif
+
+ifeq ($(CONFIG_ROME_IF),usb)
+
+ifeq ($(CONFIG_USB_PRELOAD),y)
+CDEFINES += -DCONFIG_USB_PRELOAD
+endif
+
+ifeq ($(CONFIG_FWCRASH_DUMP),y)
+CDEFINES += -DCONFIG_CRASH_DUMP_CLD
+CDEFINES += -DCONFIG_CRASH_DUMP_STACK
+CDEFINES += -DCONFIG_WLAN_FWLOG_FILE
+endif
+
+endif
+
 # Enable target dump for non-qualcomm platform
 ifeq ($(CONFIG_NON_QC_PLATFORM), y)
 CDEFINES += -DCONFIG_NON_QC_PLATFORM
-- 
1.7.9.5

