--- qcacld-new/CORE/HDD/src/wlan_hdd_cfg80211.c.orig	2015-11-04 09:46:31.130456239 +0800
+++ qcacld-new/CORE/HDD/src/wlan_hdd_cfg80211.c	2015-11-04 09:47:47.666455765 +0800
@@ -279,6 +279,10 @@
     HDD5GHZCHAN(5660,132, 0) ,
     HDD5GHZCHAN(5680,136, 0) ,
     HDD5GHZCHAN(5700,140, 0) ,
+
+/* Add 144 channel support */
+    HDD5GHZCHAN(5720,144, 0) ,
+
     HDD5GHZCHAN(5745,149, 0) ,
     HDD5GHZCHAN(5765,153, 0) ,
     HDD5GHZCHAN(5785,157, 0) ,
