--- qcacld-new/CORE/MAC/src/pe/sch/schBeaconProcess.c.orig	2015-11-27 18:43:50.661958673 +0800
+++ qcacld-new/CORE/MAC/src/pe/sch/schBeaconProcess.c	2015-11-27 18:43:58.225958626 +0800
@@ -532,6 +532,16 @@
                       pBeacon->OperatingMode.chanWidth,
                       pStaDs->staIndex, pMh->sa);
              }
+
+             /* Update Nss setting */   
+             if (pStaDs->vhtSupportedRxNss !=         
+                     (pBeacon->OperatingMode.rxNSS + 1)) {    
+                 pStaDs->vhtSupportedRxNss =              
+                     (pBeacon->OperatingMode.rxNSS + 1);      
+                 limSetNssChange( pMac, psessionEntry,    
+                         pStaDs->vhtSupportedRxNss,               
+                         pStaDs->staIndex, pMh->sa);              
+             }
           }
           else if (psessionEntry->vhtCapability && pBeacon->VHTOperation.present)
           {
