diff -Nuar qcacld-new.orig/CORE/CLD_TXRX/TXRX/ol_rx_defrag.c qcacld-new/CORE/CLD_TXRX/TXRX/ol_rx_defrag.c
--- qcacld-new.orig/CORE/CLD_TXRX/TXRX/ol_rx_defrag.c	2016-07-07 01:14:46.579842706 +0800
+++ qcacld-new/CORE/CLD_TXRX/TXRX/ol_rx_defrag.c	2016-07-07 01:15:20.011843980 +0800
@@ -282,8 +282,8 @@
     more_frag = mac_hdr->i_fc[1] & IEEE80211_FC1_MORE_FRAG;
 
     if ((!more_frag) && (!fragno) && (!rx_reorder_array_elem->head)) {
-        rx_reorder_array_elem->head = frag;
-        rx_reorder_array_elem->tail = frag;
+        ol_rx_fraglist_insert(htt_pdev, &rx_reorder_array_elem->head,
+            &rx_reorder_array_elem->tail, frag, &all_frag_present);
         adf_nbuf_set_next(frag, NULL);
         ol_rx_defrag(pdev, peer, tid, rx_reorder_array_elem->head);
         rx_reorder_array_elem->head = NULL;
