diff -Nuar qcacld-new.orig/CORE/MAC/src/pe/lim/limProcessMlmRspMessages.c qcacld-new/CORE/MAC/src/pe/lim/limProcessMlmRspMessages.c
--- qcacld-new.orig/CORE/MAC/src/pe/lim/limProcessMlmRspMessages.c	2016-06-20 11:59:22.789093306 +0800
+++ qcacld-new/CORE/MAC/src/pe/lim/limProcessMlmRspMessages.c	2016-06-21 10:02:16.767851541 +0800
@@ -430,8 +430,7 @@
 
 
         if ( (eLIM_STA_IN_IBSS_ROLE == psessionEntry->limSystemRole) ||
-             ((eLIM_AP_ROLE == psessionEntry->limSystemRole) &&
-             (vos_nv_getChannelEnabledState(channelId) != NV_CHANNEL_DFS)) )
+             ((eLIM_AP_ROLE == psessionEntry->limSystemRole)))
         {
             //Configure beacon and send beacons to HAL
             limSendBeaconInd(pMac, psessionEntry);
diff -Nuar qcacld-new.orig/firmware_bin/WCNSS_qcom_cfg.usb.ini qcacld-new/firmware_bin/WCNSS_qcom_cfg.usb.ini
--- qcacld-new.orig/firmware_bin/WCNSS_qcom_cfg.usb.ini	2016-06-20 11:56:48.925087959 +0800
+++ qcacld-new/firmware_bin/WCNSS_qcom_cfg.usb.ini	2016-06-21 10:13:31.199874977 +0800
@@ -354,6 +354,7 @@
 #If set to 0, will not scan DFS channels
 
 gEnableDFSChnlScan=1
+gEnableDFSMasterCap=1
 
 
 gVhtChannelWidth=2
