diff --git Android.mk Android.mk
index 754103f..0da77df 100644
--- Android.mk
+++ Android.mk
@@ -79,6 +79,8 @@ $(shell mkdir -p $(TARGET_OUT)/lib/modules; \
     ln -sf /system/lib/modules/$(WLAN_CHIPSET)/$(WLAN_CHIPSET)_wlan.ko \
            $(TARGET_OUT)/lib/modules/wlan.ko)
 $(shell ln -sf /persist/wlan_mac.bin $(TARGET_OUT_ETC)/firmware/wlan/qca_cld/wlan_mac.bin)
+$(shell ln -sf /persist/factory/wlan/2g_scpc_cal.bin $(TARGET_OUT_ETC)/firmware/wlan/qca_cld/2g_scpc_cal.bin)
+$(shell ln -sf /persist/factory/wlan/5g_scpc_cal.bin $(TARGET_OUT_ETC)/firmware/wlan/qca_cld/5g_scpc_cal.bin)
 
 ifeq ($(call is-board-platform-in-list, msm8960),true)
 $(shell ln -sf /firmware/image/bdwlan20.bin $(TARGET_OUT_ETC)/firmware/fakeboar.bin)
diff --git CORE/SERVICES/BMI/ol_fw.c CORE/SERVICES/BMI/ol_fw.c
index 66d741e..39c5fd8 100644
--- CORE/SERVICES/BMI/ol_fw.c
+++ CORE/SERVICES/BMI/ol_fw.c
@@ -363,6 +363,20 @@ exit:
 	return status;
 }
 
+/**
+ * qca6174_compute_checksum_only() - compute checksum of 6320 eeprom memory in uint16_t
+ * @ half_ptr: pointer to 6320 eeprom memory in uint16_t pointer
+ * @ length: length of 6320 eeprom memory in unit of uint16_t
+ *
+ * Return: sum calculated for checksum
+ */
+static u_int16_t qca6174_compute_checksum_only(u_int16_t *half_ptr, u_int16_t length)
+{
+   u_int16_t sum = 0, i;
+   for(i = 0; i < length; i++) { sum ^= *half_ptr++; }
+   return(sum);
+}
+
 static int ol_transfer_bin_file(struct ol_softc *scn, ATH_BIN_FILE file,
 			 u_int32_t address, bool compressed)
 {
@@ -542,6 +556,54 @@ static int ol_transfer_bin_file(struct ol_softc *scn, ATH_BIN_FILE file,
 
 		OS_MEMCPY(tempEeprom, (u_int8_t *)fw_entry->data, fw_entry_size);
 
+		/* BDF is copied in tempEeprom.
+		 * Check if 2g_scpc_cal.bin and 5g_scpc_cal.bin are available.
+		 * If yes, merge them into BDF and change checksum
+		 * +++ start +++
+		 */
+#define AR6320_EEPROM_CHECKSUM_OFFSET                 0x2
+#define AR6320_EEPROM_CHECKSUM_LENGTH                 0x2
+#define AR6320_EEPROM_2G_SCPC_CAL_DATA_OFFSET         0x280
+#define AR6320_EEPROM_2G_SCPC_CAL_DATA_LENGTH         0x17A
+#define AR6320_EEPROM_5G_SCPC_CAL_DATA_OFFSET         0xAE8
+#define AR6320_EEPROM_5G_SCPC_CAL_DATA_LENGTH         0x360
+		{
+			u_int16_t sum;
+			const char *filename_2g_cal = "wlan/qca_cld/2g_scpc_cal.bin";
+			const char *filename_5g_cal = "wlan/qca_cld/5g_scpc_cal.bin";
+			const struct firmware *fw_entry_2g_cal = NULL;
+			const struct firmware *fw_entry_5g_cal = NULL;
+			/* merge to BDF only if 2G cal and
+			 * 5G cal files are available
+			 */
+			if ((request_firmware(&fw_entry_2g_cal, filename_2g_cal,
+					      scn->sc_osdev->device) == 0) &&
+			    (request_firmware(&fw_entry_5g_cal, filename_5g_cal,
+					      scn->sc_osdev->device) == 0)) {
+				pr_info("%s: cal files are found\n", __func__);
+				OS_MEMCPY(tempEeprom + AR6320_EEPROM_2G_SCPC_CAL_DATA_OFFSET,
+					  (u_int8_t *)fw_entry_2g_cal->data,
+					  AR6320_EEPROM_2G_SCPC_CAL_DATA_LENGTH);
+				OS_MEMCPY(tempEeprom + AR6320_EEPROM_5G_SCPC_CAL_DATA_OFFSET,
+					  (u_int8_t *)fw_entry_5g_cal->data,
+					  AR6320_EEPROM_5G_SCPC_CAL_DATA_LENGTH);
+
+				/* compute checksum with new content */
+				OS_MEMSET(tempEeprom + AR6320_EEPROM_CHECKSUM_OFFSET,
+					  0, AR6320_EEPROM_CHECKSUM_LENGTH);
+				sum = qca6174_compute_checksum_only((u_int16_t *)tempEeprom, fw_entry_size/sizeof(u_int16_t));
+				sum = 0xFFFF ^ sum;
+				OS_MEMCPY(tempEeprom + AR6320_EEPROM_CHECKSUM_OFFSET,
+					  &sum, AR6320_EEPROM_CHECKSUM_LENGTH);
+			}
+			/* release the memory */
+			if (fw_entry_2g_cal)
+				release_firmware(fw_entry_2g_cal);
+			if (fw_entry_5g_cal)
+				release_firmware(fw_entry_5g_cal);
+		}
+		/* --- end --- */
+
 		switch (scn->target_type) {
 		default:
 			board_data_size = 0;
